# Mark That as False

## A python Project  

This project is used to distinguish which media news is true and which one is false.

1. Importing the required libraries and packages

In [1]:
import numpy as np
import pandas as pd
import itertools
import seaborn as sns
import matplotlib.pyplot as plt

Matplotlib is building the font cache; this may take a moment.


2. Reading the data from the csv file

In [13]:
df= pd.read_csv('report.csv')

3. Now, I will describe the dataset and preprocess it

In [10]:
print(df.shape)
df.head()

(6335, 4)


Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [4]:
df.describe

<bound method NDFrame.describe of       Unnamed: 0                                              title  \
0           8476                       You Can Smell Hillary’s Fear   
1          10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2           3608        Kerry to go to Paris in gesture of sympathy   
3          10142  Bernie supporters on Twitter erupt in anger ag...   
4            875   The Battle of New York: Why This Primary Matters   
...          ...                                                ...   
6330        4490  State Department says it can't find emails fro...   
6331        8062  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332        8622  Anti-Trump Protesters Are Tools of the Oligarc...   
6333        4021  In Ethiopia, Obama seeks progress on peace, se...   
6334        4330  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fe

In [5]:
df.info

<bound method DataFrame.info of       Unnamed: 0                                              title  \
0           8476                       You Can Smell Hillary’s Fear   
1          10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2           3608        Kerry to go to Paris in gesture of sympathy   
3          10142  Bernie supporters on Twitter erupt in anger ag...   
4            875   The Battle of New York: Why This Primary Matters   
...          ...                                                ...   
6330        4490  State Department says it can't find emails fro...   
6331        8062  The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...   
6332        8622  Anti-Trump Protesters Are Tools of the Oligarc...   
6333        4021  In Ethiopia, Obama seeks progress on peace, se...   
6334        4330  Jeb Bush Is Suddenly Attacking Trump. Here's W...   

                                                   text label  
0     Daniel Greenfield, a Shillman Journalism Fell

4. The label column of the dataset is the one which tells us whether the news is REAL or FAKE. So, I am taking the label column into a new variable called 'Labels'. I will be using this data to further train and test my model.

In [6]:
labels = df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

5. Now, I will be splitting the data into Train and Test sets to train and check the accuracy of my predictions

In [14]:
from sklearn.model_selection import train_test_split

In [15]:
x_train, x_test, y_train, y_test = train_test_split(df.text, labels, test_size=0.2, random_state =7)

6. Let’s initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7. Stop words are the most common words in a language that will be filtered out before processing the natural language data and the TfidfVectorizer will turn the collection of raw documents into a matrix of TF-IDF features.

I will fit and transform the vectorizer on the train set, and transform the vectorizer on the test set.

In [16]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [17]:
tfidf_vectorizer= TfidfVectorizer(stop_words = 'english', max_df =0.7)

In [18]:
tfidf_train = tfidf_vectorizer.fit_transform(x_train)
tfidf_test = tfidf_vectorizer.transform(x_test)

7. The classification of news can be carried out using many different classification algorithms.

I will be using the Passive-Aggressive Classifier for this project as it remains passive for a correct classification outcome and turns aggressive in the event of a miscalculation, updating and adjusting. Unlike most other algorithms, it does not converge. It can be implemented very easily using the Scikit-Learn Library.

In [19]:
from sklearn.linear_model import PassiveAggressiveClassifier

In [20]:
pac = PassiveAggressiveClassifier(max_iter = 50)

In [21]:
pac.fit(tfidf_train, y_train)

PassiveAggressiveClassifier(max_iter=50)

In [22]:
y_pred = pac.predict(tfidf_test)

8. Now, I will measure the accuracy score of my model which will tell us how effective and accurate this particular algorithm has been.

In [23]:
from sklearn.metrics import accuracy_score

In [24]:
score = accuracy_score(y_test,y_pred)

In [25]:
score_percent = round(score*100, 2)
print(f'{score_percent}%')

92.9%


## The accuracy percentage of this Fake News Classifier is 92.9%

9. As the last step of my project, I will be making a confusion matrix to analyse and compare the given and the predicted values and finally save the predictions as required.

In [26]:
from sklearn.metrics import confusion_matrix

In [27]:
Matrix = confusion_matrix(y_test, y_pred, labels=['FAKE', 'REAL'])

In [28]:
print(Matrix)

[[589  49]
 [ 41 588]]


So, my model gives 589 true positives, 588 true negatives, 49 false positives, and 41 false negatives.